appvia/kore
# Cascading delete and dependecy checks for teams, clusters, namespaces and service-related objects

## What

All listed changes were done only on teams, namespaces and service-related objects.

### New error type: validation.ErrDependencyViolation

This error should be used when an action can not be completed because of some existing/missing dependencies.

If you throw this error in the API logic, the API will return 409 Conflict.

Both the UI and the CLI will handle and display this error correctly. (On the UI we handle this in some cases explicitely)

### Delete options in the API: kore.DeleteOptions

With delete options you can control how a delete must be done:
 - kore.DeleteOptionSkipCheck: the delete checks must be skipped
 - kore.DeleteOptionIgnoreReadOnly: delete read-only objects
 - kore.DeleteOptionCascade: cascading delete, all child objects will be deleted (through `metadata.OwnerReferences`) with propagationPolicy=Foreground

Through some API endpoints (team, cluster, namespace, service related ones) you can set the `cascade` option to trigger a cascading delete.

If cascade=true is not set, we will check whether the object has still dependencies. In case it has, we will return with a dependency violation error (409)

See https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/ for more details.

### Delete check logic in separate methods

Pre-delete checks were moved to a DeleteCheck() method which will also be called from the relevant controllers to ensure users don't delete objects with kubectl or by other means.

### The UI doesn't do dependency checks anymore and relies on the API instead

In some cases we did some manual checks on the UI, all those were replied with the dependency violation error check.

### Uniform list filtering methods on all service objects

You can now pass filtering functions to all .List() methods.

### When creating an application service with a new namespace, we will create the namespace in the API instead of the controller

This will allow us to refresh the namespaces tab with up-to-date list after a successful service create API call. (But this is not part of this PR, see https://github.com/appvia/kore/pull/968).

## Testing

- Please create some clusters, namespaces, services, application services and service creds
- Try delete a team, cluster, etc. which has dependencies. Both the CLI and UI should throw an error
 - Try the cascading delete on different levels, it should delete all child objects successfully
